import sys
# ğŸŒŸ ë³€ê²½: QtGui ëª¨ë“ˆì„ ì¶”ê°€í•˜ì—¬ QTextCursor ìƒìˆ˜ ì ‘ê·¼ ë¬¸ì œë¥¼ í•´ê²°í•©ë‹ˆë‹¤.
from PyQt6 import QtWidgets, uic, QtCore, QtGui 
from google import genai
from google.genai.errors import APIError
from google.genai import types
# ğŸŒŸ ì¶”ê°€: í™˜ê²½ ë³€ìˆ˜ì—ì„œ API í‚¤ë¥¼ ë¶ˆëŸ¬ì˜¤ê¸° ìœ„í•´ os ëª¨ë“ˆì„ ê°€ì ¸ì˜µë‹ˆë‹¤.
import os

# =================================================================
# 1. ì„¤ì • ë° ì´ˆê¸°í™”
# =================================================================

# 1. UI íŒŒì¼ ì´ë¦„ ì„¤ì •
UI_FILE_NAME = "Mygemini.ui"

# ğŸŒŸ ì¤‘ìš”: ì‚¬ìš©ì ìš”ì²­ì— ë”°ë¼ ìƒˆë¡œìš´ API í‚¤ë¥¼ ì„ì‹œë¡œ í•˜ë“œì½”ë”©í•©ë‹ˆë‹¤. ğŸŒŸ
# ì´ í‚¤ëŠ” ì±„íŒ…ì— ë…¸ì¶œë˜ì–´ ë³´ì•ˆ ìœ„í—˜ì´ ìˆìœ¼ë©°, Googleì— ì˜í•´ ê³§ ì°¨ë‹¨ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤. 
# ìƒˆë¡œìš´ í‚¤ë¥¼ ë°œê¸‰ë°›ìœ¼ì‹  í›„, ì•„ë˜ ê°’ì„ ìƒˆë¡œìš´ í‚¤ë¡œ êµì²´í•´ ì£¼ì„¸ìš”.
GEMINI_API_KEY = "AIzaSyBViPGxOy1juy8dSEAJDJsuf-nPfwEir3o"

# 2. Gemini í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”
if not GEMINI_API_KEY:
    # í‚¤ê°€ ì—†ìœ¼ë©´ ì‹¤í–‰ì„ ì¤‘ë‹¨í•©ë‹ˆë‹¤.
    print("ì¹˜ëª…ì  ì˜¤ë¥˜: GEMINI_API_KEYê°€ ì„¤ì •ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤. ìœ íš¨í•œ í‚¤ë¥¼ ì½”ë“œì— ì…ë ¥í•´ì£¼ì„¸ìš”.")
    sys.exit(1)
    
try:
    # API í‚¤ë¥¼ genai.Client()ì— ì§ì ‘ ì „ë‹¬í•©ë‹ˆë‹¤.
    client = genai.Client(api_key=GEMINI_API_KEY)
    MODEL_NAME = "gemini-2.5-flash" # ì‚¬ìš©í•  ëª¨ë¸ ì§€ì •
except ValueError as e:
    # ì´ ë¶€ë¶„ì€ ë³´í†µ API í‚¤ í˜•ì‹ì´ ì˜ëª»ë˜ì—ˆì„ ë•Œ ë°œìƒí•©ë‹ˆë‹¤.
    print(f"í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì˜¤ë¥˜: API í‚¤ê°€ ì˜ëª»ë˜ì—ˆê±°ë‚˜ ëˆ„ë½ë˜ì—ˆìŠµë‹ˆë‹¤. {e}")
    sys.exit(1)


# =================================================================
# 2. Gemini API í˜¸ì¶œì„ ìœ„í•œ ì›Œì»¤ ìŠ¤ë ˆë“œ (QThread)
# =================================================================

class GeminiWorker(QtCore.QThread):
    # ê²°ê³¼ ë° ì˜¤ë¥˜ë¥¼ ë©”ì¸ ìŠ¤ë ˆë“œë¡œ ì „ë‹¬í•˜ê¸° ìœ„í•œ ì‹œê·¸ë„ ì •ì˜
    # response_ready(user_question, gemini_response)
    response_ready = QtCore.pyqtSignal(str, str) 
    # error_occurred(error_type, error_message)
    error_occurred = QtCore.pyqtSignal(str, str) 

    def __init__(self, client, model_name, chat_history, user_question):
        super().__init__()
        self.client = client
        self.model_name = model_name
        # ëŒ€í™” ê¸°ë¡ì€ ì°¸ì¡°ë¡œ ì „ë‹¬ë˜ì–´, ìŠ¤ë ˆë“œ ë‚´ì—ì„œ ì—…ë°ì´íŠ¸ë©ë‹ˆë‹¤.
        self.chat_history = chat_history 
        self.user_question = user_question
        
        # types.GenerateContentConfig ê°ì²´ ìƒì„±
        self.config = types.GenerateContentConfig(
            system_instruction="You are a helpful assistant. Please answer all questions in Korean."
        )

    def run(self):
        """API í˜¸ì¶œì„ ë³„ë„ì˜ ìŠ¤ë ˆë“œì—ì„œ ì‹¤í–‰í•˜ì—¬ GUI ë©ˆì¶¤ì„ ë°©ì§€í•©ë‹ˆë‹¤."""
        try:
            # API í˜¸ì¶œ: ì „ì²´ ëŒ€í™” ê¸°ë¡ì„ ì „ë‹¬í•©ë‹ˆë‹¤. (ë„¤íŠ¸ì›Œí¬ ë¸”ë¡œí‚¹ ë°œìƒ ì§€ì )
            response = self.client.models.generate_content(
                model=self.model_name,
                contents=self.chat_history,
                config=self.config
            )
            
            gemini_response = response.text
            
            # ğŸŒŸ ëŒ€í™” ê¸°ë¡ ì—…ë°ì´íŠ¸: API ì‘ë‹µ ì„±ê³µ ì‹œì—ë§Œ ëª¨ë¸ ë©”ì‹œì§€ë¥¼ chat_historyì— ì¶”ê°€í•©ë‹ˆë‹¤.
            self.chat_history.append({"role": "model", "parts": [{"text": gemini_response}]})

            # ì‘ë‹µ ì„±ê³µ ì‹œ ë©”ì¸ ìŠ¤ë ˆë“œë¡œ ê²°ê³¼ ì‹ í˜¸ ë°œìƒ
            self.response_ready.emit(self.user_question, gemini_response)

        except APIError as e:
            self.error_occurred.emit("API ì˜¤ë¥˜", f"Gemini ì„œë²„ì— ì—°ê²°í•  ìˆ˜ ì—†ìŠµë‹ˆë‹¤. (ì˜¤ë¥˜: {e})")
        except Exception as e:
            self.error_occurred.emit("ì˜ˆê¸°ì¹˜ ì•Šì€ ì˜¤ë¥˜", str(e))


# =================================================================
# 3. ë©”ì¸ ì• í”Œë¦¬ì¼€ì´ì…˜ í´ë˜ìŠ¤
# =================================================================

class GeminiApp(QtWidgets.QDialog):
    def __init__(self):
        super().__init__()
        
        try:
            uic.loadUi(UI_FILE_NAME, self)
        except FileNotFoundError:
            print(f"ì˜¤ë¥˜: {UI_FILE_NAME} íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤. ê²½ë¡œë¥¼ í™•ì¸í•´ ì£¼ì„¸ìš”.")
            sys.exit(1)
            
        self.setWindowTitle(f"Gemini Q&A - ëª¨ë¸: {MODEL_NAME}")
        
        # 3. ìœ„ì ¯ ì°¾ê¸° (ì´ì „ ì½”ë“œì™€ ë™ì¼)
        self.btnSend = self.findChild(QtWidgets.QPushButton, 'btnSend')
        self.lineEditMyQuestion = self.findChild(QtWidgets.QLineEdit, 'lineEditMyQuestion')
        
        self.lblAnswer = self.findChild(QtWidgets.QTextEdit, 'lblAnswer') 
        if not self.lblAnswer:
            self.lblAnswer = self.findChild(QtWidgets.QPlainTextEdit, 'lblAnswer')
        
        if not self.btnSend or not self.lblAnswer or not self.lineEditMyQuestion:
            missing = []
            if not self.btnSend: missing.append("'btnSend' (QPushButton)")
            if not self.lblAnswer: missing.append("'lblAnswer' (QTextEdit ë˜ëŠ” QPlainTextEdit ì´ì–´ì•¼ í•©ë‹ˆë‹¤!)")
            if not self.lineEditMyQuestion: missing.append("'lineEditMyQuestion' (QLineEdit)")
            
            print(f"ì¹˜ëª…ì  ì˜¤ë¥˜: UI íŒŒì¼ ({UI_FILE_NAME})ì—ì„œ ë‹¤ìŒ í•„ìˆ˜ ìœ„ì ¯ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤: {', '.join(missing)}")
            sys.exit(1)

        # 4. ìœ„ì ¯ ì´ë²¤íŠ¸ ì—°ê²° ë° ì„¤ì •
        self.btnSend.clicked.connect(self.generate_response)
        self.lblAnswer.setReadOnly(True)

        # ğŸŒŸ ëŒ€í™” ê¸°ë¡ì„ ì €ì¥í•  ë¦¬ìŠ¤íŠ¸ë¥¼ ì´ˆê¸°í™”í•©ë‹ˆë‹¤.
        self.chat_history = []
        # ğŸŒŸ ì›Œì»¤ ìŠ¤ë ˆë“œ ì¸ìŠ¤í„´ìŠ¤ë¥¼ ì €ì¥í•  ë³€ìˆ˜ë¥¼ ì´ˆê¸°í™”í•©ë‹ˆë‹¤.
        self.gemini_worker = None 
        
        self.lblAnswer.setText("[Mygemini] ë¬´ì—‡ì„ ë„ì™€ë“œë¦´ê¹Œìš”?")

        self.show()

    def generate_response(self):
        """ì‚¬ìš©ì ì§ˆë¬¸ì„ ì²˜ë¦¬í•˜ê³  ì›Œì»¤ ìŠ¤ë ˆë“œë¥¼ ì‹œì‘í•˜ì—¬ APIë¥¼ í˜¸ì¶œí•©ë‹ˆë‹¤."""
        
        user_question = self.lineEditMyQuestion.text().strip()
        
        if not user_question:
            return
        
        # 5a. ì‚¬ìš©ì ì§ˆë¬¸ì„ í™”ë©´ì— í‘œì‹œ
        user_message = f"[ì§ˆë¬¸] {user_question}\n"
        self.lblAnswer.append(user_message)
        
        # 5b. ì‘ë‹µ ìƒì„± ì¤‘ì„ì„ ì‚¬ìš©ìì—ê²Œ í‘œì‹œí•˜ê³  UI ì—…ë°ì´íŠ¸
        loading_message = "[Mygemini] ì‘ë‹µì„ ìƒì„±í•˜ëŠ” ì¤‘ì…ë‹ˆë‹¤..."
        self.lblAnswer.append(loading_message)
        QtWidgets.QApplication.processEvents() 
        
        # 5c. ëŒ€í™” ê¸°ë¡ì— ì‚¬ìš©ì ë©”ì‹œì§€ë¥¼ ì¶”ê°€í•©ë‹ˆë‹¤ (API í˜¸ì¶œì„ ìœ„í•´)
        self.chat_history.append({"role": "user", "parts": [{"text": user_question}]})

        # 6. Gemini API í˜¸ì¶œ (QThread ì‚¬ìš©)
        self.gemini_worker = GeminiWorker(
            client=client,
            model_name=MODEL_NAME,
            chat_history=self.chat_history,
            user_question=user_question
        )
        
        # ì›Œì»¤ ìŠ¤ë ˆë“œì˜ ì‹œê·¸ë„ì„ ë©”ì¸ ìŠ¤ë ˆë“œì˜ ìŠ¬ë¡¯(ë©”ì„œë“œ)ì— ì—°ê²°
        self.gemini_worker.response_ready.connect(self.handle_response)
        self.gemini_worker.error_occurred.connect(self.handle_error)
        
        # ìŠ¤ë ˆë“œ ì‹œì‘ (GUI ë©ˆì¶¤ ë°©ì§€)
        self.gemini_worker.start()
        
        # 9. QLineEdit ë‚´ìš© ì§€ìš°ê¸°
        self.lineEditMyQuestion.clear()
        
    def handle_response(self, user_question, gemini_response):
        """API ì‘ë‹µì„ ë°›ì•„ UIì— ìµœì¢…ì ìœ¼ë¡œ í‘œì‹œí•˜ê³  ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤."""
        
        # 7. ì§€ì •ëœ í˜•ì‹ìœ¼ë¡œ ì¶œë ¥ í…ìŠ¤íŠ¸ í¬ë§·íŒ…
        formatted_output = f"[Mygemini] {gemini_response}\n"
        
        # 8. QTextEdit/QPlainTextEditì— í…ìŠ¤íŠ¸ ì„¤ì • (ìŠ¤í¬ë¡¤ ì§€ì›)
        # ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•˜ê³  ìµœì¢… ì‘ë‹µì„ ì¶”ê°€í•˜ëŠ” ê³¼ì •:
        
        # 8a. ë§ˆì§€ë§‰ ì¤„ (ë¡œë”© ë©”ì‹œì§€)ì„ ì œê±°í•©ë‹ˆë‹¤.
        cursor = self.lblAnswer.textCursor()
        # ğŸŒŸ ìˆ˜ì •: QTextCursorëŠ” QtGui ëª¨ë“ˆì— ìˆìŠµë‹ˆë‹¤.
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        # 8b. ìµœì¢… ì‘ë‹µì„ ì¶”ê°€í•©ë‹ˆë‹¤.
        self.lblAnswer.append(formatted_output)

        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()
        
    def handle_error(self, error_type, error_message):
        """API ì˜¤ë¥˜ ë°œìƒ ì‹œ UIì— ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ í‘œì‹œí•˜ê³  ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤."""
        
        # ì˜¤ë¥˜ ë°œìƒ ì‹œ ëŒ€í™” ê¸°ë¡ì—ì„œ ë§ˆì§€ë§‰ ì‚¬ìš©ì ì§ˆë¬¸ í•­ëª©ì„ ì œê±°í•©ë‹ˆë‹¤. 
        # (ì˜¤ë¥˜ë¡œ ì¸í•´ API í˜¸ì¶œì´ ì‹¤íŒ¨í–ˆìœ¼ë¯€ë¡œ ì‘ë‹µì´ ì—†ì—ˆê¸° ë•Œë¬¸)
        if self.chat_history and self.chat_history[-1].get("role") == "user":
             self.chat_history.pop()

        # 7. ì˜¤ë¥˜ ë©”ì‹œì§€ í¬ë§·íŒ…
        formatted_output = f"[Mygemini] {error_type} ë°œìƒ: {error_message}\n"
        
        # 8. ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•˜ê³  ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ ì¶”ê°€í•˜ëŠ” ê³¼ì •:
        cursor = self.lblAnswer.textCursor()
        # ğŸŒŸ ìˆ˜ì •: QTextCursorëŠ” QtGui ëª¨ë“ˆì— ìˆìŠµë‹ˆë‹¤.
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        # ìµœì¢… ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ ì¶”ê°€í•©ë‹ˆë‹¤.
        self.lblAnswer.append(formatted_output)

        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()


# =================================================================
# 4. í”„ë¡œê·¸ë¨ ì‹¤í–‰
# =================================================================

if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    window = GeminiApp()
    sys.exit(app.exec())